pinot-core/src/main/java/com/linkedin/pinot/core/operator/MCombineGroupByOperator.java

```diff
@@ -138,35 +140,34 @@ public Block getNextBlock() {
    */
   private IntermediateResultsBlock combineBlocks()
       throws InterruptedException {
-    final int numOperators = _operators.size();
-    final IntermediateResultsBlock[] blocks = new IntermediateResultsBlock[numOperators];
+    int numOperators = _operators.size();
+    final IntermediateResultsBlock[] intermediateResultsBlocks = new IntermediateResultsBlock[numOperators];
     final CountDownLatch operatorLatch = new CountDownLatch(numOperators);
-
-    final List<AggregationInfo> aggregationsInfo = _brokerRequest.getAggregationsInfo();
-    final int numAggrFunctions = aggregationsInfo.size();
-
-    // TODO: use new aggregation functions to combine results.
-    final List<AggregationFunction> aggregationFunctions =
-        AggregationFunctionFactory.getAggregationFunction(_brokerRequest);
-
     final Map<String, Serializable[]> resultsMap = new ConcurrentHashMap<>();
 
+    List<AggregationInfo> aggregationInfos = _brokerRequest.getAggregationsInfo();
+    final AggregationFunctionContext[] aggregationFunctionContexts =
+        AggregationFunctionUtils.getAggregationFunctionContexts(aggregationInfos, null);
```

Comment:
The second argument is the segment metadata. On per segment operator, we use segment metadata to construct a visitor and visit each aggregation function (visitor mode).